### PR TITLE
[QUIZ-582] subtask update quizzes UI with skip button

### PR DESCRIPTION
**Pull request by v-i-s-h-n-u-ps**
This PR adds the ability to skip questions that are marked as skippable.

INDEX_KEY: `key_wJSdZSiesX5hiVLt`

A couple of questions in the `coffee-quiz` are marked skippable.

https://app.constructor.io/dashboard/quizzes/edit/coffee-quiz?back=%2Fdashboard%2Fquizzes%3Findex_key%3Dkey_wJSdZSiesX5hiVLt

### CODE REVIEW

File: src/hooks/useQuizEvents/useQuizSkipClick.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import { useCallback } from 'react';
+import { ActionAnswerQuestion, QuestionTypes } from '../../components/CioQuiz/actions';
+import { QuizAPIReducerState } from '../../components/CioQuiz/quizApiReducer';
+import { QuizLocalReducerState } from '../../components/CioQuiz/quizLocalReducer';
+import { QuizEventsReturn, OnQuizSkipQuestion } from '../../types';
+import { isFunction } from '../../utils';
+
+const useQuizNextClick = (
```
Comment (mocca102) [inline src/hooks/useQuizEvents/useQuizSkipClick.ts:8]: I believe you meant `useQuizSkipClick` here

File: src/hooks/useQuizEvents/useQuizSkipClick.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  quizApiState: QuizAPIReducerState,
+  quizLocalState: QuizLocalReducerState,
+  dispatchLocalState: React.Dispatch<ActionAnswerQuestion>,
+  onQuizSkipQuestion?: OnQuizSkipQuestion
+): QuizEventsReturn.NextQuestion => {
+  const quizNexClickHandler = useCallback(() => {
```
Comment (mocca102) [inline src/hooks/useQuizEvents/useQuizSkipClick.ts:14]: I believe you meant `quizSkipClickHandler`

Comment (v-i-s-h-n-u-ps) [reply src/hooks/useQuizEvents/useQuizSkipClick.ts]: [39da1d4](https://github.com/Constructor-io/constructorio-ui-quizzes/pull/132/commits/39da1d4e2a01f97440c535a87b5777047d6a8499)

File: src/hooks/useQuizEvents/useQuizSkipClick.ts (hunk 1 above)
Comment (v-i-s-h-n-u-ps) [reply src/hooks/useQuizEvents/useQuizSkipClick.ts]: [39da1d4](https://github.com/Constructor-io/constructorio-ui-quizzes/pull/132/commits/39da1d4e2a01f97440c535a87b5777047d6a8499)

**Review by mocca102 — APPROVED** [review]
LGTM! lets get a second review as well

File: src/components/CioQuiz/quizLocalReducer.ts [hunk 3]
```diff
@@ -105,11 +105,7 @@ export default function quizLocalReducer(
           newAnswers.push(['seen']);
           break;
         case QuestionTypes.SingleSelect:
-          newAnswers.push([]);
```
Comment (mocca102) [inline src/components/CioQuiz/quizLocalReducer.ts]: nice refactor

File: src/components/CioQuiz/quizLocalReducer.ts [hunk 4]
```diff
@@ -92,6 +92,31 @@ export default function quizLocalReducer(
         isQuizCompleted: false,
       };
     }
+    case QuestionTypes.Skip: {
+      const { answers, answerInputs } = state;
+      const newAnswers = [...answers];
+      const lastAnswerInputIndex = answers.length;
+      const currentAnswerInput = Object.values(state.answerInputs)?.[lastAnswerInputIndex];
+      switch (currentAnswerInput.type) {
```
Comment (esezen) [inline src/components/CioQuiz/quizLocalReducer.ts:100]: The previous line has optional chaining, does that mean `currentAnswerInput` could be null/undefined? If yes, is it possible for currentAnswerInput.type to error out? I just realized it's the same in `Next` so maybe it's a question for @mocca102 as well

File: src/components/CioQuiz/quizLocalReducer.ts [hunk 5]
```diff
@@ -92,6 +92,31 @@ export default function quizLocalReducer(
...
+          break;
+        case QuestionTypes.SingleSelect:
+        case QuestionTypes.MultipleSelect:
+        default:
+          newAnswers.push([]);
+      }
+      return {
+        ...state,
+        // We now commit current answers to prevAnswerInputs
+        prevAnswerInputs: answerInputs,
```
Comment (esezen) [inline src/components/CioQuiz/quizLocalReducer.ts:115]: I'm not super familiar with this file's logic but should we be writing to `prevAnswerInputs` here?

File: src/components/CioQuiz/quizLocalReducer.ts [hunk 6]
```diff
@@ -92,6 +92,31 @@ export default function quizLocalReducer(
...
+        answers: newAnswers,
```
Comment (esezen) [inline src/components/CioQuiz/quizLocalReducer.ts:116]: Do we need to write `answerInputs` here?

File: src/stories/Quiz/tests/mocks.ts [hunk 7]
```diff
@@ -205,6 +205,10 @@ export const useMockContextValue = (question?: Question): QuizContextValue => {
       ...mockElementProps,
       className: 'cio-question-cta-button',
     }),
+    getSkipQuestionButtonProps: () => ({
+      ...mockElementProps,
+      className: 'cio-question-cta-button',
```
Comment (esezen) [inline src/stories/Quiz/tests/mocks.ts:210]: Should this be `cio-question-skip-button` or does it not matter?

**Review by esezen — COMMENTED** [review]
Thanks for working on this, this is looking great!

File: src/components/CioQuiz/quizLocalReducer.ts (hunk 4 above)
Comment (mocca102) [reply src/components/CioQuiz/quizLocalReducer.ts:100]: The optional chaining actually should be removed.

Comment (mocca102) [reply src/components/CioQuiz/quizLocalReducer.ts:100]: It's not doing anything but it's throwing whomever reading the code off

File: src/stories/Quiz/tests/mocks.ts (hunk 7 above)
Comment (v-i-s-h-n-u-ps) [reply src/stories/Quiz/tests/mocks.ts]: [f18aa78](https://github.com/Constructor-io/constructorio-ui-quizzes/pull/132/commits/f18aa78821419475ac9b701bcd44f4f165819fca)

File: src/components/CioQuiz/quizLocalReducer.ts (hunk 4 above)
Comment (v-i-s-h-n-u-ps) [reply src/components/CioQuiz/quizLocalReducer.ts:100]: Removed [f18aa78](https://github.com/Constructor-io/constructorio-ui-quizzes/pull/132/commits/f18aa78821419475ac9b701bcd44f4f165819fca)

File: src/components/CioQuiz/quizLocalReducer.ts (hunk 5 above)
Comment (esezen) [reply src/components/CioQuiz/quizLocalReducer.ts:115]: @mocca102 thoughts?

File: src/components/CioQuiz/quizLocalReducer.ts (hunk 6 above)
Comment (esezen) [reply src/components/CioQuiz/quizLocalReducer.ts:116]: @mocca102 thoughts?

File: src/components/CioQuiz/quizLocalReducer.ts (hunk 5 above)
Comment (mocca102) [reply src/components/CioQuiz/quizLocalReducer.ts:115]: What do you mean by? @esezen 

>  writing to prevAnswerInputs here?

**Review by esezen — APPROVED** [review]
LGTM!

**Pull request merged**